core/classes/class-query.php

```diff
@@ -115,6 +115,65 @@ public function get_past_events( int $number = 5 ): \WP_Query {
 	public function pre_get_posts( $query ) {
 		$events_query = $query->get( 'gp_events_query' );
 
+		if ( ! is_admin() && $query->is_main_query() ) {
+			$general = get_option( Utility::prefix_key( 'general' ) );
+			$pages   = $general['pages'];
+
+			if ( empty( $pages ) || ! is_array( $pages ) ) {
+				return;
+			}
+
+			$archive_pages = array(
+				'past'     => json_decode( $pages['past_events'] ),
+				'upcoming' => json_decode( $pages['upcoming_events'] ),
+			);
+
+			foreach ( $archive_pages as $key => $value ) {
+				if ( ! empty( $value ) && is_array( $value ) ) {
+					$page = $value[0];
+					if ( $page->id === $query->queried_object_id ) {
+						$query->set( 'post_type', 'gp_event' );
+
+						$page_id                     = $query->queried_object_id;
+						$events_query                = $key;
+						$query->is_page              = false;
+						$query->is_singular          = false;
+						$query->is_archive           = true;
+						$query->is_post_type_archive = array( Event::POST_TYPE );
+
+						// This will force a page to behave like an archive page. Use -1 as that is not a valid ID.
+						$query->queried_object_id = '-1';
```

Comment:
This line and `pre_option` filter below is required due to this conditional in `get_posts`:

https://github.com/WordPress/wordpress-develop/blob/6.1/src/wp-includes/class-wp-query.php#L2120